camply/providers/recreation_dot_gov/recdotgov_permits.py

```diff
@@ -0,0 +1,393 @@
+"""
+Recreation.gov Implementation for Tours.
+"""
+
+import json
+import logging
+import functools
+from abc import ABC, abstractmethod
+from datetime import date, datetime, time, timedelta, timezone
+from typing import Any, Dict, List, Optional, Tuple
+
+import pandas as pd
+import requests
+
+from camply.config import RecreationBookingConfig, RIDBConfig
+from camply.containers import AvailableCampsite
+from camply.containers.api_responses import (
+    RecDotGovSearchResponse,
+    RecDotGovSearchResult,
+    RecDotGovPermitMapping,
+    RecDotGovPermitMappingResponse,
+    PermitEntranceResponse,
+    PermitEntranceForFacilityResponse,
+    PermitMonthlyAvailabilityResponse,
+)
+from camply.containers.base_container import CamplyModel, RecDotGovEquipment
+from camply.providers.base_provider import ProviderSearchError
+from camply.providers.recreation_dot_gov.recdotgov_provider import RecreationDotGovBase
+from camply.utils import api_utils
+
+logger = logging.getLogger(__name__)
+
+
+class RecreationDotGovPermit(RecreationDotGovBase):
+    """
+    Recreation.gov Implementation for Permits
+    """
+
+    api_response_class = PermitEntranceResponse
+    api_search_result_class = RecDotGovSearchResult
+    api_search_result_key = "entity_id"
+    api_search_fq = "entity_type:permit"
+    api_base_path = RecreationBookingConfig.API_BASE_PATH
+    facility_type = RIDBConfig.PERMIT_FACILITY_FIELD_QUALIFIER
+    activity_name = None  # Activity Name Should't Be Propogated to Query Parameters
+    resource_api_path = RIDBConfig.PERMIT_ENTRANCE_API_PATH
+    booking_url = "https://www.recreation.gov/permits/{facility_id}/registration/detailed-availability"
+
+    facility_permit_map = {
+        # Yosemite National Park Wilderness Permits
+        445859: {
+            44585904: "Beehive Meadows",
+            44585907: "Cathedral Lakes",
+            44585914: "Glen Aulin",
+            44585915: "Glen Aulin -> Cold Canyon/Waterwheel (pass through)",
+            44585917: "Happy Isles -> Little Yosemite Valley (No Donohue Pass)",
+            44585918: "Happy Isles -> Past LYV (Donohue Pass Eligible)",
+            44585921: "Lyell Canyon (Donohue Pass Eligible)",
+            44585924: "May Lake",
+            44585940: "Rafferty Creek -> Vogelsang",
+            44585945: "Sunrise",
+            44585956: "Young Lakes via Glen Trail",
+        }
+    }
```

Comment:
Is there any way to rely on the Recreation.gov API for this? Camply has a built-in API key to talk to the RIDB API here: https://ridb.recreation.gov/docs#/Permit%20Entrances/getFacilityPermitEntrances

Reply:
@juftin looking through #22 comments, there are multiple end points  for permit search. Would searching through each of them for a given query be a good idea or would it be out of scope for Camply?

Reply:
If possible, I don't want to hit every endpoint. It looks like which endpoint to use is actually handled here: https://github.com/changlan/camply/blob/e5425eebd811256d04d7108bd786a439fa11e8fb/camply/providers/recreation_dot_gov/recdotgov_permits.py#L66-L96

It seems like everyone uses the same generic endpoint, unless they show up on this mapping: https://www.recreation.gov/api/permitcontent/permitmapping

Reply:
So the workflow for this seems to be, 
1. Call the permitmapping API.
2. For each facility in the permitmapping result, check if it is a parent facility.
3. If it has children, create a subdictionary with its children, and use the facility retrieval API to get each child's name.
4. If not, input it in the mapping dictionary with its name as the value.

I think I will take a stab at it.